node/service/src/lib.rs

```diff
@@ -416,28 +490,88 @@ pub fn new_full<RuntimeApi, Executor>(
 		})
 		.collect();
 
-	let (overseer, overseer_handler) = real_overseer(leaves, prometheus_registry.as_ref(), spawner)?;
-	let overseer_handler_clone = overseer_handler.clone();
+	let authority_discovery_service = {
+		let mut ads = None;
+
+		if matches!(role, Role::Authority{..} | Role::Sentry{..}) {
+			use sc_network::Event;
+			use futures::StreamExt;
+
+			if !authority_discovery_disabled {
+				let (sentries, authority_discovery_role) = match role {
+					Role::Authority { ref sentry_nodes } => (
+						sentry_nodes.clone(),
+						authority_discovery::Role::Authority (
+								keystore_container.keystore(),
+						),
+					),
+					Role::Sentry {..} => (
+						vec![],
+						authority_discovery::Role::Sentry,
+					),
+					_ => unreachable!("Due to outer matches! constraint; qed."),
+				};
+
+				let network_event_stream = network.event_stream("authority-discovery");
+				let dht_event_stream = network_event_stream.filter_map(|e| async move { match e {
+						Event::Dht(e) => Some(e),
+						_ => None,
+				}});
+				let (authority_discovery_worker, service) = authority_discovery::new_worker_and_service(
+						client.clone(),
+						network.clone(),
+						sentries,
+						Box::pin(dht_event_stream),
+						authority_discovery_role,
+						prometheus_registry.clone(),
+				);
+
+				task_manager.spawn_handle().spawn("authority-discovery-worker", authority_discovery_worker.run());
+				ads = Some(service);
+			}
+		}
 
-	task_manager.spawn_essential_handle().spawn_blocking("overseer", Box::pin(async move {
-		use futures::{pin_mut, select, FutureExt};
+		ads
+	};
 
-		let forward = polkadot_overseer::forward_events(overseer_client, overseer_handler_clone);
+	let overseer_handler = match authority_discovery_service {
```

Comment:
Wait, why don't we start the overseer when authority discovery is disabled?

Can we not just register a dummy authority discovery subsystem?

Reply:
I was just asking about that in Element. The authority discovery service is necessary to initialize the NetworkBridge subsystem, and it seemed plausible that the overseer wouldn't be able to do the right thing without the NetworkBridge anyway.

Reply:
This makes the authority discovery a required feature for authority nodes. Which is okay, but than a validator shouldn't be able to start with authority discovery disabled.

Reply:
Yeah, it is required now. We could make it an error in cli in a separate PR.

Reply:
> @mxinden what is the reason authority discovery is still "optional"?

The authority discovery module was not mandatory thus far. The flag would allow a smoother feature roll out. I.e. first put it behind a feature flag then enable it by default but allow to disable it and eventually remove the possibility of disabling it.

Given that with the introduction of the _overseer_ the authority discovery module becomes mandatory on authority nodes, I would suggest removing the flag entirely, always starting the authority discovery module on validator and sentry nodes. (The latter being deprecated https://github.com/paritytech/substrate/issues/6845.)

Would that solve the issue at hand?

Reply:
@coriolinus friendly ping. What do you think of my suggestion above?

Reply:
I certainly don't mind the idea of removing the flag entirely! I'd prefer to do it in a distinct PR, though; this one has enough going on, and people suggesting that it should be burned in, that I'd really like to keep the changeset as small as possible for now. I'll create an issue to remove the flag, though.

Reply:
https://github.com/paritytech/polkadot/issues/1848

Reply:
Sounds good to me. Thanks.